### PR TITLE
update fld training with regard to given reference

### PR DESCRIPTION
**Pull request by rayeaster**
This change reflect the reference paper implementation.

http://alumni.cs.ucr.edu/~hli/paper/hli05tumor.pdf

Test result on demo data set shows the change has a better error rate with a smooth decision boundary. Check related issues [here](https://github.com/haifengl/smile/issues/320)

image

old demo error rate
`training error = 21.00% for toy and training error = 20.14% for big toy`

new demo error rate
`training error = 19.00% for toy and training error = 18.73% for big toy`

image

image

### CODE REVIEW

**Comment by haifengl** [comment]
Thanks. What about other datasets, e.g. unit tests? The demo is only for, well, "demo". They are not serious test cases.

File: core/src/main/java/smile/classification/FLD.java [hunk 1]
```diff
@@ -273,32 +278,51 @@ public FLD(double[][] x, int[] y, int L, double tol) {
...
-        }
-
+        }     
+        //
+        // calculate the square root of diagonal inverse of eigenvalue matrix from SVD of covariance matrix
+        //
         DenseMatrix U = eigen.getEigenVectors();
-        DenseMatrix UB = U.atbmm(B);
-
+        DenseMatrix D = eigen.getD().inverse();
```
Comment (haifengl) [inline core/src/main/java/smile/classification/FLD.java]: This is slow and also not numerically stable.  D is a diagonal matrix, which may have zeros in the diagonal. But `inverse()` is a general method, suitable for dense matrix.

File: core/src/main/java/smile/classification/FLD.java [hunk 2]
```diff
@@ -273,32 +278,51 @@ public FLD(double[][] x, int[] y, int L, double tol) {
...
+        EVD eigen = T.eigen();  
         
         tol = tol * tol;
         double[] s = eigen.getEigenValues();
         for (int i = 0; i < s.length; i++) {
             if (s[i] < tol) {
                 throw new IllegalArgumentException("The covariance matrix is close to singular.");
             }
-
-            s[i] = 1.0 / s[i];
```
Comment (haifengl) [inline core/src/main/java/smile/classification/FLD.java]: why remove this?

**Comment by rayeaster** [comment]
Good point . I could try more dataset or any perticular dataset you have on mind

File: core/src/main/java/smile/classification/FLD.java (hunk 1 above)
Comment (rayeaster) [reply core/src/main/java/smile/classification/FLD.java]: This is from the reference paper if what I understand is correct.
image

File: core/src/main/java/smile/classification/FLD.java (hunk 2 above)
Comment (rayeaster) [reply core/src/main/java/smile/classification/FLD.java]: not quite clear about the usage of this line

**Comment by rayeaster** [comment]
And I think current LDA implementation is actually PCA style

File: core/src/test/java/smile/classification/NaiveBayesTest.java [hunk 3]
```diff
@@ -266,7 +344,7 @@ public void testLearnBernoulli() {
             }
         }
 
-        System.out.format("Bernoulli error = %d of %d%n", error, total);
```
Comment (haifengl) [inline core/src/test/java/smile/classification/NaiveBayesTest.java]: `%n` is a portal way for `\n`. Please don't change this.

File: core/src/test/java/smile/classification/NaiveBayesTest.java [hunk 4]
```diff
@@ -307,7 +385,7 @@ public void testLearnBernoulli2() {
...
```
Comment (haifengl) [inline core/src/test/java/smile/classification/NaiveBayesTest.java]: same above.

File: core/src/test/java/smile/classification/NaiveBayesTest.java [hunk 5]
```diff
@@ -101,7 +101,7 @@ public void tearDown() {
      */
```
Comment (haifengl) [inline core/src/test/java/smile/classification/NaiveBayesTest.java]: Can we separate this file change from the above? They are not related topics.

**Comment by haifengl** [comment]
There is no `U.abmm(U.atbmm(B))`. You missed the lines 291-295. We actually calculate inv(S_t) * S_b.

File: core/src/test/java/smile/classification/NaiveBayesTest.java (hunk 4 above)
Comment (rayeaster) [reply core/src/test/java/smile/classification/NaiveBayesTest.java]: Yes I will keep it as it was. Thanks for pointing out

**Comment by haifengl** [comment]
It is very expensive and instable to calculate inv(T) directly. In general, it is not a good idea to translate your math formula to code directly. Computational mathematics studies how to calculate them quickly and reliably. It is the gap you missed.

**Comment by haifengl** [comment]
For all the data you tested, we should get the same results with the old implementation and new one because the data do not in the small sample size problem as discussed in the paper. The difference may be caused by the numerical stability. Iris has only 4 variables. So it is probably better that we use the old implementation that works on T (4x4 matrix, very small). For other data with high dimensions, we should use the fast algorithm/new implementation.

Of course, the difference may also caused by some bug :) I will take some time to review both implementations. Thanks!

**Comment by rayeaster** [comment]
So which do you prefer? Keep current implementation as it is or you want to see more tests before considering new implementation:)

**Comment by rayeaster** [comment]
That sounds great:) Thank you for taking time to review this. Maybe we could use a parmeter to switch between 2 implementations for different data size. Also a lot thanks for the comment about computational math.BTW any resource on that topic you recommend as reference?

**Comment by haifengl** [comment]
Yes, it is what I would like.

**Comment by rayeaster** [comment]
Great. I could add the parameter

**Comment by rayeaster** [comment]
@haifengl do you have time to review the implementation difference. Thanks.

**Comment by rayeaster** [comment]
@haifengl please let me know if you see any implementation mistake in this Pull Request or you spot any inappropriate coding style. Thanks.

> 10-fold

oss

**Comment by haifengl** [comment]
I will cherry pick your PR. I still believe that two algorithms should generate the same model for the test data in the unit tests. These data don't have the small sample size problem. So the two algorithms should be same mathematically.

**Comment by rayeaster** [comment]
@haifengl maybe because the old implementation calculate `inv(S_t) * S_b`, not `inv(S_w) * S_b`?

**Comment by haifengl** [comment]
Mathematically, `inv(S_t) * S_b` and `inv(S_w) * S_b` are same in sense of optimizing the FLD criterion when `S_w` is not singular.

**Comment by rayeaster** [comment]
so what should be the cause for discrepancy there?

**Comment by haifengl** [comment]
I have cherry-picked this into v2. Thanks!

**Pull request closed without merging**